### PR TITLE
Added the reference implementation POUF as POUF 1

### PR DESCRIPTION
**Pull request by mnm678**

### CODE REVIEW

File: POUF1.md [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+
+**Note that this implementation is not meant to be used in practice**. It is a guide for implementers to understand how the system should function.
+
+The reference implementation includes all required and most optional features as a reference for others looking to implement Uptane.
+
+# Protocols
+This POUF uses ASN.1/DER encoding for transmitting files. The details of the file formats are described in {Formats} in ASN.1 notation. Note that improper decoding of ASN.1 may lead to arbitrary code execution or Denial-of-Service attacks. Any implementers of this POUF should be aware of these attacks and use a decoder which has been subject to appropriate security analysis.
+
+The Primary ECU communicates with the Director repository, Image repository, and time server using http post requests over [XML-RPC](http://xmlrpc.scripting.com/spec.html). In this communication, the Primary ECU acts as an XML-RPC client and the Director repository, Image repository, and time server act as XML-RPC servers. The requests supported by these servers are described in the Message Handler Table.
```
Comment (JustinCappos) [inline POUF1.md]:
```suggestion
The Primary ECU communicates with the Director repository, Image repository, and time server using HTTP POST requests over [XML-RPC](http://xmlrpc.scripting.com/spec.html). In this communication, the Primary ECU acts as an XML-RPC client and the Director repository, Image repository, and time server act as XML-RPC servers. The requests supported by these servers are described in the Message Handler Table.
```

File: POUF1.md [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+
+This POUF does not specify communication between Primary and Secondary ECUs. This communication is left up to the implementer, but should follow the requirements of the standard. For more information about internal vehicle communication, see the [Deployment Considerations](https://uptane.github.io/deployment/introduction.html).
```
Comment (JustinCappos) [inline POUF1.md]: If this is true then how can this be interoperable?  Does it qualify as a POUF?  Why is it okay to miss this?

File: POUF1.md [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+
+Additionally, this POUF supports custom delegated targets roles and terminating delegations, features described as optional in the [Standard](https://uptane.github.io/uptane-standard/uptane-standard.html#targets_role_delegations). As explained in the Standard, terminating delegations are not used by default. To use custom roles, the Targets role on the Image repository sets up a delegation. Delegations used in this POUF must support multi-role delegations, as explained in [TAP 3](https://github.com/theupdateframework/taps/blob/d0818e580c322815a473520f2e8cc5f5eb8df499/tap3.md). In addition, delegation filenames are listed using an enumerated list that supports [wildcards](https://uptane.github.io/uptane-standard/uptane-standard.html#delegations_meta). Delegations are supported in the reference implementation to demonstrate how they can be used to establish a chain of trust for specific images.
```
Comment (JustinCappos) [inline POUF1.md]: Can wildcards be everywhere?  What wildcard characters are supported?  Can you use them in directories?  How do they match directory paths (e.g., does "a*.txt" match "apple/file.txt")?

File: POUF1.md [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+
+Public key identifiers use the hash of the key value, encryption algorithm, and verification scheme for a condensed identifier. The format of these identifiers if further described in {Formats}. This POUF supports rsa and ed25519 hashes for public keys. The hashing algorithm must be specified in the signature method.
```
Comment (JustinCappos) [inline POUF1.md]: capitalize, etc. as appropriate

File: POUF1.md [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+
+This POUF supports asymmetric keys and uses the [same keys for encrypting and signing](https://uptane.github.io/uptane-standard/uptane-standard.html#build-time-prerequisite-requirements-for-ecus). Using the same keys allows for a simplified reference implementation and less key management. In particular, each ECU has a private key and registers the associated public key with the director repository. The ECU will use the private key to sign messages and the director repository may use the public key to encrypt images per ECU. In addition, each metadata role has either online or offline private keys that are used on the director and image repositories. The public keys associated with these roles are used by the ECUs during validation.
```
Comment (JustinCappos) [inline POUF1.md]: "use the private key to sign messages"  Is this just the manifest or more?

File: POUF1.md [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+
+The Image repository and Director repository have public interfaces for communication with ECUs. To make the public interface, the Image and Directory repositories each store metadata in a filesystem that is hosted to form the public interface. These interfaces do not require authentication to read.
+
+The Director repository identifies a vehicle at the start of the update process using a stored vehicle version manifest. The Director checks the vehicle version manifest as described in the standard, but does not make any additional checks. This POUF does not do dependency resolution as the reference implementation is not used in practice and so does not have conflicts.
```
Comment (JustinCappos) [inline POUF1.md]: A few odd things here.  What if an ECU was replaced?  What do you mean about dependency resolution here?  This is for packages on a single ECU.  What does your statement have to do with conflicts.

File: POUF1.md [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+
+# Metadata
+This POUF supports some metadata features that are optional in the Uptane Standard, but are required for an implementation of this POUF. These features are included in the metadata definitions in {Formats}. The details of these features are:
+* The metadata version will increment on an update to the metadata.
+* Root metadata must support [mappings of roles to lists of URLs as described in TAP 5](https://uptane.github.io/uptane-standard/uptane-standard.html#root_meta).
+* Repository mapping metadata will be included using a map file. The format of this file is described in {Formats}. In the mapping metadata, files are expressed using an enumerated list that supports wildcards.
+* The OEM will provide unencrypted images to ECUs, hence the Director repository’s targets metadata will contain hashes of unencrypted images. Metadata about encrypted images may be included in the custom field of targets metadata as described in the [Standard](https://uptane.github.io/uptane-standard/uptane-standard.html#custom-metadata-about-images).
```
Comment (JustinCappos) [inline POUF1.md]: I thought encryption was optionally supported as was discussed above...

File: POUF1.md [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+* The custom field of targets metadata on the Director and Image repositories are not compared. This means that the must match functionality described in the [Standard](https://uptane.github.io/uptane-standard/uptane-standard.html#custom-metadata-about-images) is not supported. A client may choose to compare some fields of targets metadata, but these checks are not required by this POUF.
+* The Image repository’s targets metadata may include metadata about delegations. [Delegations may be any depth and may also be cyclic](https://uptane.github.io/uptane-standard/uptane-standard.html#delegations_meta).
+* Snapshot metadata in this POUF does not include the root filename or version.
+
+# Usage
+As the Uptane reference implementation is not used in practice, key management and image generation are not handled by this POUF. The POUF uses online keys for all roles to demonstrate Uptane functionality. This practice is not recommended for Uptane implementers.
```
Comment (JustinCappos) [inline POUF1.md]: You keep saying it isn't used in practice.  Why is this relevant?

File: POUF1.md [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+* The custom field of targets metadata on the Director repository must support an [encrypted target field for encrypting images per ECU](https://uptane.github.io/uptane-standard/uptane-standard.html#custom-metadata-about-images). In addition, the custom field will contain a release counter and hardware id from both the Director and Image repositories. This POUF does not support the director adding a download URL to the custom field of targets metadata.
+* The custom field of targets metadata on the Director and Image repositories are not compared. This means that the must match functionality described in the [Standard](https://uptane.github.io/uptane-standard/uptane-standard.html#custom-metadata-about-images) is not supported. A client may choose to compare some fields of targets metadata, but these checks are not required by this POUF.
+* The Image repository’s targets metadata may include metadata about delegations. [Delegations may be any depth and may also be cyclic](https://uptane.github.io/uptane-standard/uptane-standard.html#delegations_meta).
+* Snapshot metadata in this POUF does not include the root filename or version.
```
Comment (JustinCappos) [inline POUF1.md]: Why would it and why doesn't it?

File: POUF1.md [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+* The custom field of targets metadata on the Director repository must support an [encrypted target field for encrypting images per ECU](https://uptane.github.io/uptane-standard/uptane-standard.html#custom-metadata-about-images). In addition, the custom field will contain a release counter and hardware id from both the Director and Image repositories. This POUF does not support the director adding a download URL to the custom field of targets metadata.
+* The custom field of targets metadata on the Director and Image repositories are not compared. This means that the must match functionality described in the [Standard](https://uptane.github.io/uptane-standard/uptane-standard.html#custom-metadata-about-images) is not supported. A client may choose to compare some fields of targets metadata, but these checks are not required by this POUF.
+* The Image repository’s targets metadata may include metadata about delegations. [Delegations may be any depth and may also be cyclic](https://uptane.github.io/uptane-standard/uptane-standard.html#delegations_meta).
```
Comment (JustinCappos) [inline POUF1.md]: Aren't these always true?

File: POUF1.md [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+* The custom field of targets metadata on the Director repository must support an [encrypted target field for encrypting images per ECU](https://uptane.github.io/uptane-standard/uptane-standard.html#custom-metadata-about-images). In addition, the custom field will contain a release counter and hardware id from both the Director and Image repositories. This POUF does not support the director adding a download URL to the custom field of targets metadata.
+* The custom field of targets metadata on the Director and Image repositories are not compared. This means that the must match functionality described in the [Standard](https://uptane.github.io/uptane-standard/uptane-standard.html#custom-metadata-about-images) is not supported. A client may choose to compare some fields of targets metadata, but these checks are not required by this POUF.
```
Comment (JustinCappos) [inline POUF1.md]: I wasn't aware this is optional.  Is it?  

How are conflicts / mismatches handled?

File: POUF1.md [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+* The Image repository’s targets metadata may include metadata about delegations. [Delegations may be any depth and may also be cyclic](https://uptane.github.io/uptane-standard/uptane-standard.html#delegations_meta).
+* Snapshot metadata in this POUF does not include the root filename or version.
+
+# Usage
+As the Uptane reference implementation is not used in practice, key management and image generation are not handled by this POUF. The POUF uses online keys for all roles to demonstrate Uptane functionality. This practice is not recommended for Uptane implementers.
+
+## Server Setup
+The Uptane Reference Implementation does the following setup before any updates can be installed:
+* The Director repository is initialized with the private keys for all roles with online keys.
```
Comment (JustinCappos) [inline POUF1.md]: Why online keys?  Please explain.

File: POUF1.md (hunk 3 above)
Comment (mnm678) [reply POUF1.md]: I clarified that these are shell-style wildcards. I can add more description if needed.

File: POUF1.md (hunk 5 above)
Comment (mnm678) [reply POUF1.md]: Yes, this is just the manifest, I changed it to make that more clear.

File: POUF1.md (hunk 6 above)
Comment (mnm678) [reply POUF1.md]: I clarified this a bit and removed the reference to dependency resolution. Dependency resolution seems to be out of scope for this document as it does not affect any of the communication and the standard already requires that some form of dependency resolution is used.

File: POUF1.md (hunk 7 above)
Comment (mnm678) [reply POUF1.md]: This is referring to images that are encrypted per ECU. The Image repository does not know which ECU will receive the image, and so only contains metadata for the unencrypted image. The Director knows the ECU that will receive the image and so can add the encrypted metadata to the custom field.

I tried to make this more clear in the description.

File: POUF1.md (hunk 10 above)
Comment (mnm678) [reply POUF1.md]: I read this as a MAY in the standard (https://uptane.github.io/uptane-standard/uptane-standard.html#delegations_meta), but it does seem like something that should always be supported.

File: POUF1.md (hunk 11 above)
Comment (mnm678) [reply POUF1.md]: Yes, this is optional. All other fields of Targets metadata are compared, but the custom field is not (except when must match is implemented). If all other fields match, the custom metadata from the Director is used.

**Comment by mnm678** [comment]
@JustinCappos I updated the document to address your comments.

File: POUF1.md [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+* The Image repository’s targets metadata may include metadata about delegations. [Delegations may be any depth and may also be cyclic](https://uptane.github.io/uptane-standard/uptane-standard.html#delegations_meta).
+* Snapshot metadata in this POUF does not include the root filename or version. These fields are included in implementations that do not use [TAP 5](https://github.com/theupdateframework/taps/blob/01726d203c9b9c029d26f6612069ce3180500d9a/tap5.md#downloading-metadata-and-target-files). Because this POUF implements TAP 5, the root filename and version do not need to be included in Snapshot metadata.
+
+# Usage
+Key management and image generation are not handled by this POUF. The POUF uses online keys for all roles to demonstrate Uptane functionality. This practice is not recommended for Uptane implementers.
+
+## Server Setup
+The Uptane Reference Implementation does the following setup before any updates can be installed:
+* The Director repository is initialized with the private keys for all roles with online keys. These keys are stored on the Director so they can be used to automatically sign valid metadata.
```
Comment (JustinCappos) [inline POUF1.md]: Explain / rationalize

Comment (awwad) [reply POUF1.md]: o.o Director root definitely can't be an online key.  The demo's keys are all open, but I doubt that's what you mean?

Comment (mnm678) [reply POUF1.md]: I added a sentence to clarify that not all keys should be stored online.

**Comment by JustinCappos** [comment]
Approved by Ira.

**Pull request merged**